src/celeritas/io/RootEventReader.cc

```diff
@@ -0,0 +1,118 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/io/RootEventReader.cc
+//---------------------------------------------------------------------------//
+#include "RootEventReader.hh"
+
+#include <TFile.h>
+#include <TLeaf.h>
+#include <TTree.h>
+
+#include "corecel/io/Logger.hh"
+#include "celeritas/ext/ScopedRootErrorHandler.hh"
+#include "celeritas/phys/ParticleParams.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with ROOT input filename.
+ */
+RootEventReader::RootEventReader(std::string const& filename,
+                                 SPConstParticles params)
+    : params_(std::move(params))
+{
+    ScopedRootErrorHandler scoped_root_error;
+
+    CELER_EXPECT(!filename.empty());
+    tfile_.reset(TFile::Open(filename.c_str(), "read"));
+    CELER_ASSERT(tfile_->IsOpen());
+    ttree_.reset(tfile_->Get<TTree>(tree_name()));
+    CELER_ASSERT(ttree_);
+    num_entries_ = ttree_->GetEntries();
+    CELER_ASSERT(num_entries_ > 0);
+
+    scoped_root_error.throw_if_errors();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Read single event from the primaries tree.
+ */
+auto RootEventReader::operator()() -> result_type
+{
+    ScopedRootErrorHandler scoped_root_error;
+
+    CELER_EXPECT(entry_count_ <= num_entries_);
```

Comment:
Put preconditions at the very very top. Same with the constructor above.

```suggestion
    CELER_EXPECT(entry_count_ <= num_entries_);

    ScopedRootErrorHandler scoped_root_error;
```